### DISCUSSION TITLE
Multiple notifications

### DISCUSSION THREAD

**Discussion by paulsizer** [excid3/noticed#327] [General]
We use noticed to show the end user (coach) a list of notifications they have received from their clients.

I have an action that allows a client to complete a workout, once the workout is completed I create a notification. 

`WorkoutNotification.with(workout: @workout, sender: session_user.client).deliver_later(@workout.client.coach)`

Inside the notification I have:

` deliver_by :database, if: :should_send?`

```
def should_send?
    notification_doesnt_exist? && notification_enabled?
  end
```

```
def notification_doesnt_exist?
    workout = Workout.find(params[:workout][:id])
    coach = workout.client.coach
    client = workout.client
    coach.notifications.where(type: "WorkoutNotification", params: {workout: workout, sender: client}).blank?
  end
```

notification_enabled just checks the coaches settings to see if they allow those type of notifications.

Some users complain they are receiving multiple notifications for the same thing. I have tried to replicate this by creating a loop in the console and creating multiple notifications with the same params and it only ever creates 1.

Should I do a check outside of the WorkoutNotification class?

**Comment by matthee** [comment]
Hey @paulsizer!

I just stumbled upon your issue and I hope this is still relevant!

Your problem sounds like a race-condition to me. Nothing in your code prevents running `notification_doesnt_exist?` in parallel. For example, when these notifications are created by some sort of background job, it might be possible for two jobs to check `notification_doesnt_exist?` at the same time (before the notification is inserted to the DB). This would result in both jobs performing the insert and  the same record would get added to the database twice. 

Maybe, locking the `coach` record first resolves your issue, like so?

```ruby
around_deliver :with_locked_coach 

def with_locked_coach(&block)
  workout = Workout.find(params[:workout][:id])
  workout.client.coach.with_lock(&block)
end
```

 However, it would probably be better to perform this check before calling `deliver_*` on your notification.
